EasyCommands/CommandParsers/ParameterProcessors.cs

```diff
@@ -116,1146 +319,307 @@ public static class ParameterProcessorRegistry {
-                else if (p is GroupCommandParameter && !isGroup) isGroup = true;
-                else return false;
-                return true;
-            }
+        static DataProcessor<T> rightList<T>(bool required) where T: class, CommandParameter {
+            return new DataProcessor<T>(new ListValueDataFetcher<T>(required), df => p => false, df => p => df.SetValue(p));
         }
 
-        public class VariableSelectorProcessor : SimpleParameterProcessor<VariableSelectorCommandParameter> {
-            BlockType? blockType = null;
-            bool isGroup = false;
-
-            public override bool CanConvert(List<CommandParameter> p) {
-                return blockType != null;
-            }
-
-            public override CommandParameter Convert(List<CommandParameter> p) {
-                VariableSelectorCommandParameter selector = findFirst<VariableSelectorCommandParameter>(p);
-                Variable variable = selector.Value;
-                return new SelectorCommandParameter(new SelectorEntityProvider(blockType.Value, isGroup, variable));
-            }
-
-            public override void Initialize() {
-                blockType = null;
-                isGroup = false;
-            }
-
-            public override bool ProcessLeft(CommandParameter p) {
-                return false;
-            }
-
-            public override bool ProcessRight(CommandParameter p) {
-                if (p is BlockTypeCommandParameter && blockType == null) blockType = ((BlockTypeCommandParameter)p).Value;
-                else if (p is GroupCommandParameter && !isGroup) isGroup = true;
-                else return false;
-                return true;
-            }
+        //ParameterProcessors
+        public interface ParameterProcessor {
+            List<Type> GetProcessedTypes();
+            bool CanProcess(CommandParameter p); 
+            bool Process(List<CommandParameter> p, int i, out List<CommandParameter> finalParameters);
+            int GetRank();
+            void SetRank(int rank);
         }
 
-        public class PrintCommandProcessor : SimpleParameterProcessor<PrintCommandParameter> {
-            Variable variable;
-
-            public override bool CanConvert(List<CommandParameter> p) {
-                return variable != null;
-            }
-
-            public override CommandParameter Convert(List<CommandParameter> p) {
-                return new CommandReferenceParameter(new PrintCommand(variable));
-            }
-
-            public override void Initialize() {
-                variable = null;
+        public abstract class ParameterProcessor<T> : ParameterProcessor where T : class, CommandParameter {
+            int rank;
+            public virtual List<Type> GetProcessedTypes() { return new List<Type>() { typeof(T) }; }
+            public virtual int GetRank() => rank;
+            public virtual void SetRank(int r) => rank = r;
```

Comment:
Pretty sure `Rank` can be a property. `public virtual int Rank { get; set; }`

Reply:
can't use virtual on the interface, but i was able to condense it down pretty far! :) Basically have to copy the definition (plus public) in the abstract class